Phenopacket2

pnrobinson opened the pull request:
See new ClinicalTextMapperTest

@tudorgroza trying to fix this

 20-03-2020 -- instead of getting dolichocephalic head shape and broad nasal bridge and narrow palate,
Fenominal was applying "narrow" to "broad nasal bridge" and inferred Narrow nasal bridge
Note -- with the version of the HP we are using for testing, dolichocephalic head shape is not an exact match
However, broad nasal bridge is but it does not match

tudorgroza commented:
@pnrobinson - There are 2 issues in the way we do tokenization:

1. the pattern matcher used in SimpleSentence to compile the tokens is too restrictive and unfortunately has a second 'hidden' role of "cleaning" the tokens. For example (from the sentence used in the test), " ... intellectual impairment (IQ 65), macrocephaly ..." ends up being tokenised as ["intellectual", "impairment", "IQ", "macrocephaly"]. As you can see, "65)" is gone, together with all the punctuation. This happens also in the case of " ... broad nasal bridge, narrow palate ..." which leads to ["broad", "nasal", "bridge", "narrow", "palate"] which then leads to [bridge narrow] being considered a candidate when matching 2-token entries.

2. the stop word list removes conjunction markers - e.g. _and_, which leads to: " ... macrocephaly and dysmorphisms ..." being tokenised as ["macrocephaly", "dysmorphisms"]

There's no easy fix. I can rewrite the tokeniser to take into account punctuation and conjunctions. Would suggest to do that in a different branch / ticket. Let me know.

pnrobinson commented:
@tudorgroza I added a new classes called OptimalSentenceMapper that solves the above problem and is still pretty fast. I added various interfacces that should make it easier to further optimize. Let's discuss what needs to be done by zoom.
